spec/nmt.md

```diff
@@ -0,0 +1,303 @@
+// and with the IgnoreMaxNamespace set to true
+tree := New(sha256.New(), NamespaceIDSize(1), InitialCapacity(4), IgnoreMaxNamespace(true))
+```
+
+One can examine the namespace ID size of the `tree` using
+
+```go
+func (n *NamespacedMerkleTree) NamespaceSize() namespace.IDSize
+```
+E.g.,
+```go
+idSize := tree.NamespaceSize() // outputs 1
+```
+### Ignore Max Namespace 
+If the NMT is configured with `IgnoreMaxNamespace` set to true, then the calculation of the namespace ID range of non-leaf nodes in the [namespace hash function](#namespaced-hash) will change slightly.
+That is, when determining the upper limit of the namespace ID range for a tree node `n`, with children `l` and `r`, the maximum possible namespace ID, equivalent to `NamespaceIDSize()` bytes of `0xFF`, or `2^NamespaceIDSize()-1`,
+should be omitted if feasible (in the preceding code example with the ID size of `1` byte, the maximum possible namespace ID would be `0xFF`). 
+This is achieved by taking the maximum value among the namespace IDs available in the range of node's left and right children (i.e., `n.maxNs = max(l.minNs, l.maxNs , r.minNs, r.maxNs))`, which is not equal to the maximum possible namespace ID value. 
+If such a namespace ID does not exist, the `maxNs` is calculated as normal, i.e., `n.maxNs = max(l.maxNs , r.maxNs)`.
+
+## Add Leaves
+
+Data items are added to the tree using the `Push` method.
+Data items should be prefixed with namespaces of size set out for the NMT (i.e.,   `tree.NamespaceSize()`) and added in ascending order of their namespace IDs to avoid errors during the `Push` process. 
+Non-compliance with either of these requirements cause `Push` to fail.
+
+```go
+func (n *NamespacedMerkleTree) Push(namespacedData namespace.PrefixedData) error
+```
+E.g.,
+```go
+d := append(namespace.ID{0}, []byte("leaf_0")...) // the first `tree.NamespaceSize()` bytes of each data item is treated as its namespace ID.
+if err := tree.Push(d); err != nil {
+	// something went wrong
+}
+// add a few more data items
+d1 := append(namespace.ID{0}, []byte("leaf_1")...) 
+if err := tree.Push(d1); err != nil {
+    // something went wrong
+}
+d2 := append(namespace.ID{1}, []byte("leaf_2")...) 
+if err := tree.Push(d2); err != nil {
+    // something went wrong
+}
+d3 := append(namespace.ID{3}, []byte("leaf_3")...) 
+if err := tree.Push(d3); err != nil {
+    // something went wrong
+}
+```
+## Get Root
+
+The `Root()` method calculates the NMT root based on the data that has been added through the use of the `Push` method.
+```go
+func (n *NamespacedMerkleTree) Root() []byte
+```
+For example:
+```go
+// compute the root
+root := tree.Root() 
+```
+In the provided code example, the root would be `00 03 b1c2cc5` (as also illustrated in Figure 1).
+
+The minimum and maximum namespace IDs of the tree root can be obtained through the following methods:
+
+```go
+minNS := nmt.MinNamespace(root, tree.NamespaceSize())
+maxNS := nmt.MaxNamespace(root, tree.NamespaceSize())
```

Comment:
[no change needed] I wonder if these methods should be added because they seem slightly more conveinet:

```go
minNS := tree.MinNamespace() 
maxNS := tree.MaxNamespace() 
```

given the tree struct already has exported methods for `tree.Root()` and `tree.NamespaceSize()`, this seems feasible.

Reply:
Sure, noted https://github.com/celestiaorg/nmt/issues/104